src/features/envVisualizer/components/Arrow.tsx

```diff
@@ -0,0 +1,118 @@
+import { KonvaEventObject } from 'konva/types/Node';
+import { Arrow as KonvaArrow } from 'react-konva';
+
+import { Config } from '../EnvVisualizerConfig';
+import { Layout } from '../EnvVisualizerLayout';
+import { Hoverable, Visible } from '../EnvVisualizerTypes';
+import { setHoveredStyle, setUnhoveredStyle } from '../EnvVisualizerUtils';
+import { ArrayUnit } from './binding/value/ArrayUnit';
+import { ArrayValue } from './binding/value/ArrayValue';
+import { FnValue } from './binding/value/FnValue';
+import { GlobalFnValue } from './binding/value/GlobalFnValue';
+import { Frame } from './Frame';
+import { Text } from './Text';
+
+/** this class encapsulates an arrow to be drawn between 2 points */
+export class Arrow implements Visible, Hoverable {
+  readonly x: number;
+  readonly y: number;
+  readonly height: number;
+  readonly width: number;
+  readonly points: number[];
+
+  constructor(readonly from: Visible, readonly to: Visible) {
+    this.x = from.x;
+    this.y = from.y;
+
+    if (from instanceof Frame) {
```

Comment:
Would it be better to define a "from coordinate" and "to coordinate" for arrow drawing that is overridden within each subclass of Visible, instead of accounting for all of them here?

Reply:
Thanks for your suggestion, this should be possible. However, one concern may be that we should not add too many fields to the classes and have it become harder to understand and maintain. 2 extra coordinates fields just for drawing of arrows added to each class may be unnecessarily adding more fields and having logic meant for drawing arrows be split into those classes. But I do see where you are coming from as well, since the coordinates to draw from and to are specific to the classes themselves. What are your thoughts? If you feel strongly about this I'll try to change it.

Reply:
Hmm. But I think drawing arrows is sort of a "core" function of the program since there will be many instances where this happens in more complicated structures. So I do see an advantage in moving the arrow logic into each of those classes.

Reply:
alright, will change it up. thanks for the suggestion!

Reply:
addressed [here](https://github.com/source-academy/cadet-frontend/pull/1676/commits/c5d80c1f9a7fb9f54f8a89deeef2d229ebf1587b)